File: src/main/java/com/matag/game/cardinstance/CardInstanceFactory.java
package com.matag.game.cardinstance;

import com.matag.cards.Card;
import com.matag.cards.properties.Rarity;
import com.matag.game.status.GameStatus;
import org.springframework.beans.BeansException;
import org.springframework.context.ApplicationContext;
import org.springframework.context.ApplicationContextAware;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.TreeSet;
import java.util.stream.Collectors;

import static java.util.Collections.emptyList;

@Component
public class CardInstanceFactory implements ApplicationContextAware {
  private ApplicationContext applicationContext;

  @Override
  public void setApplicationContext(ApplicationContext applicationContext) throws BeansException {
    this.applicationContext = applicationContext;
  }

  public CardInstance create(GameStatus gameStatus, int id, Card card, String owner, String controller) {
    var cardInstance = applicationContext.getBean(CardInstance.class);
    cardInstance.setGameStatus(gameStatus);
    cardInstance.setId(id);
    cardInstance.setCard(card);
    cardInstance.setOwner(owner);
    cardInstance.setController(controller);
    return cardInstance;
  }

  public CardInstance create(GameStatus gameStatus, int id, Card card, String owner) {
    return create(gameStatus, id, card, owner, null);
  }

  public CardInstance mask(CardInstance cardInstance) {
    return create(cardInstance.getGameStatus(), cardInstance.getId(), hiddenCard(), cardInstance.getOwner());
  }

  public List<CardInstance> mask(List<CardInstance> cardInstanceList) {
    return cardInstanceList.stream()
      .map(this::mask)
      .collect(Collectors.toList());
  }

  private Card hiddenCard() {
    return new Card("card", "/img/card-back.jpg", new TreeSet<>(), emptyList(), new TreeSet<>(), new TreeSet<>(), Rarity.COMMON, "", 0, 0, emptyList());
  }
}

File: src/main/kotlin/com/matag/game/cardinstance/CardInstanceFactory.kt
package com.matag.game.cardinstance

import com.matag.cards.Card
import com.matag.cards.ability.Ability
import com.matag.cards.properties.*
import com.matag.game.status.GameStatus
import org.springframework.beans.BeansException
import org.springframework.context.ApplicationContext
import org.springframework.context.ApplicationContextAware
import org.springframework.stereotype.Component
import java.util.*
import java.util.stream.Collectors

@Component
class CardInstanceFactory : ApplicationContextAware {
    private var applicationContext: ApplicationContext? = null

    @Throws(BeansException::class)
    override fun setApplicationContext(applicationContext: ApplicationContext) {
        this.applicationContext = applicationContext
    }

    @JvmOverloads
    fun create(
        gameStatus: GameStatus?,
        id: Int,
        card: Card?,
        owner: String?,
        controller: String? = null
    ): CardInstance {
        val cardInstance = applicationContext!!.getBean<CardInstance>(CardInstance::class.java)
        cardInstance.gameStatus = gameStatus
        cardInstance.id = id
        cardInstance.card = card
        cardInstance.owner = owner
        cardInstance.controller = controller
        return cardInstance
    }

    fun mask(cardInstance: CardInstance): CardInstance {
        return create(cardInstance.gameStatus, cardInstance.id, hiddenCard(), cardInstance.owner)
    }

    fun mask(cardInstanceList: List<CardInstance>): List<CardInstance> {
        return cardInstanceList
            .map { cardInstance: CardInstance? -> this.mask(cardInstance!!) }
    }

    private fun hiddenCard(): Card {
        return Card(
            "card",
            "/img/card-back.jpg",
            TreeSet(),
            listOf(),
            TreeSet(),
            TreeSet(),
            Rarity.COMMON,
            "",
            0,
            0,
            listOf()
        )
    }
}
